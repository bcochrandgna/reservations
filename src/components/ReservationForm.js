// @flow
import React, { PureComponent } from 'react';
import { View, Text, TextInput } from 'react-native';
import { Button } from 'react-native-elements';
import moment from 'moment';
import DatePicker from 'react-native-datepicker';
import styles from '../styles/components/ReservationForm';
import { Reservation } from '../types/Reservation';
import i18n from '../helpers/i18n';

type Props = {
  onUpdate: (reservation: Reservation) => Promise<void>,
};
type State = {
  name: string,
  hotelName: string,
  arrivalDate: string,
  departureDate: string,
};

class ReservationForm extends PureComponent<Props, State> {
  state = {
    name: '',
    hotelName: '',
    arrivalDate: moment()
      .add(1, 'days')
      .format('MM/DD/YYYY'),
    departureDate: moment()
      .add(2, 'days')
      .format('MM/DD/YYYY'),
  };

  handleSavePress = () => {
    const {
      name, hotelName, arrivalDate, departureDate,
    } = this.state;
    const { onUpdate } = this.props;

    onUpdate({
      id: null,
      name,
      hotelName,
      arrivalDate,
      departureDate,
    });
  };

  handleTextChange = (field: string, text: string) => {
    this.setState({
      [field]: text,
    });
  };

  render() {
    const {
      name, hotelName, arrivalDate, departureDate,
    } = this.state;

    return (
      <View style={styles.container}>
        <Text>Name</Text>
        <TextInput
          style={styles.inputContainer}
          value={name}
          onChangeText={text => this.handleTextChange('name', text)}
        />
        <Text>Hotel Name</Text>
        <TextInput
          style={styles.inputContainer}
          value={hotelName}
          onChangeText={text => this.handleTextChange('hotelName', text)}
        />
        <Text>Arrival Date</Text>
        <DatePicker
          format="MM/DD/YYYY"
          confirmBtnText="Confirm"
          cancelBtnText="Cancel"
          date={arrivalDate}
          mode="date"
          showIcon={false}
          onDateChange={date => this.handleTextChange('arrivalDate', date)}
          style={styles.dateContainer}
          customStyles={{
            dateInput: styles.dateInput,
          }}
        />
        <Text>Departure Date</Text>
        <DatePicker
          format="MM/DD/YYYY"
          confirmBtnText="Confirm"
          cancelBtnText="Cancel"
          date={departureDate}
          mode="date"
          showIcon={false}
          onDateChange={date => this.handleTextChange('departureDate', date)}
          style={styles.dateContainer}
          customStyles={{
            dateInput: styles.dateInput,
          }}
        />
        <Button title={i18n.t('create_reservation')} onPress={this.handleSavePress} />
      </View>
    );
  }
}

export default ReservationForm;
