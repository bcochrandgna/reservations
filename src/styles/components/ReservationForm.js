import { StyleSheet } from 'react-native';
import { standardPadding, standardMargin, borderColor } from '..';

export default StyleSheet.create({
  container: {
    padding: standardPadding,
    margin: standardMargin,
  },
  inputContainer: {
    borderBottomWidth: 1,
    borderBottomColor: borderColor,
    marginVertical: standardMargin,
    padding: standardPadding,
    paddingBottom: standardPadding * 2,
  },
  dateContainer: {
    margin: 0,
    marginBottom: standardMargin,
    width: '100%',
  },
  dateInput: {
    margin: 0,
    borderWidth: 0,
    borderBottomWidth: 1,
    borderBottomColor: borderColor,
    alignItems: 'flex-start',
    padding: standardPadding,
  },
});
