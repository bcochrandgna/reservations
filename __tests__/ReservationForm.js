import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import ReservationForm from '../src/components/ReservationForm';

jest.mock('DatePickerIOS', () => 'DatePickerIOS');

const handleUpdate = jest.fn();

describe('ReservationForm', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<ReservationForm />);
    expect(wrapper).toMatchSnapshot();
  });

  it('submits properly', () => {
    const instance = renderer.create(<ReservationForm onUpdate={handleUpdate} />).getInstance();
    instance.handleTextChange('name', 'Name 1');
    instance.handleTextChange('hotelName', 'Hotel Name 1');
    instance.handleTextChange('arrivalDate', '03/01/2019');
    instance.handleTextChange('departureDate', '03/03/2019');
    instance.handleSavePress();

    expect(handleUpdate).toHaveBeenCalled();
    expect(handleUpdate).toHaveBeenCalledWith({
      id: null,
      name: 'Name 1',
      hotelName: 'Hotel Name 1',
      arrivalDate: '03/01/2019',
      departureDate: '03/03/2019',
    });
  });
});
