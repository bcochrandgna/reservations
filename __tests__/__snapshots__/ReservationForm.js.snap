// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReservationForm renders correctly 1`] = `
<View
  style={
    Object {
      "margin": 5,
      "padding": 5,
    }
  }
>
  <Text>
    Name
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    style={
      Object {
        "borderBottomColor": "#dddddd",
        "borderBottomWidth": 1,
        "marginVertical": 5,
        "padding": 5,
        "paddingBottom": 10,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Text>
    Hotel Name
  </Text>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    style={
      Object {
        "borderBottomColor": "#dddddd",
        "borderBottomWidth": 1,
        "marginVertical": 5,
        "padding": 5,
        "paddingBottom": 10,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <Text>
    Arrival Date
  </Text>
  <DatePicker
    TouchableComponent={[Function]}
    allowFontScaling={true}
    androidMode="default"
    cancelBtnText="Cancel"
    confirmBtnText="Confirm"
    customStyles={
      Object {
        "dateInput": Object {
          "alignItems": "flex-start",
          "borderBottomColor": "#dddddd",
          "borderBottomWidth": 1,
          "borderWidth": 0,
          "margin": 0,
          "padding": 5,
        },
      }
    }
    date="02/28/2019"
    disabled={false}
    duration={300}
    format="MM/DD/YYYY"
    height={259}
    hideText={false}
    iconSource={
      Object {
        "testUri": "../../react-native-datepicker/date_icon.png",
      }
    }
    modalOnResponderTerminationRequest={[Function]}
    mode="date"
    onDateChange={[Function]}
    placeholder=""
    showIcon={false}
    style={
      Object {
        "margin": 0,
        "marginBottom": 5,
        "width": "100%",
      }
    }
  />
  <Text>
    Departure Date
  </Text>
  <DatePicker
    TouchableComponent={[Function]}
    allowFontScaling={true}
    androidMode="default"
    cancelBtnText="Cancel"
    confirmBtnText="Confirm"
    customStyles={
      Object {
        "dateInput": Object {
          "alignItems": "flex-start",
          "borderBottomColor": "#dddddd",
          "borderBottomWidth": 1,
          "borderWidth": 0,
          "margin": 0,
          "padding": 5,
        },
      }
    }
    date="03/01/2019"
    disabled={false}
    duration={300}
    format="MM/DD/YYYY"
    height={259}
    hideText={false}
    iconSource={
      Object {
        "testUri": "../../react-native-datepicker/date_icon.png",
      }
    }
    modalOnResponderTerminationRequest={[Function]}
    mode="date"
    onDateChange={[Function]}
    placeholder=""
    showIcon={false}
    style={
      Object {
        "margin": 0,
        "marginBottom": 5,
        "width": "100%",
      }
    }
  />
  <ForwardRef(Themed.Button)
    onPress={[Function]}
    title="Create Reservation"
  />
</View>
`;
